RavenDB-22469 Read entry IDs (actual values) in bulk when performing deep paging.

maciejaszyk opened the pull request:
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-24152
https://issues.hibernatingrhinos.com/issue/RavenDB-22469
### Additional description

Read IDs required for proper paging in bulk instead of one-by-one.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [x] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-22469 Read entry IDs (actual values) in bulk when performing deep paging.

ayende commented on src/Corax/Querying/TermsReader.cs:
```diff
@@ -41,6 +55,74 @@ public string GetTermFor(long id)
+                var containers = Vector128.Load(ptr);
+                Vector128.Divide(containers, Vector128.Create((long)Voron.Global.Constants.Storage.PageSize)).Store(ptr);
+            }
+        }
```
You are missing `else` instruction here. If you have 512, you probably also have 256 and 128 - so this is executed multiple times.

Why are you doing a _divide_ instead of shift, isn't that much cheaper?

maciejaszyk replied:
a) Good point, in worst case we will have to process up to 8 elements scalarly. I've added `else`
b) Readability. I changed it into right shift.

ayende commented on src/Corax/Querying/TermsReader.cs:
```diff
@@ -41,6 +55,74 @@ public string GetTermFor(long id)
+        _lookup.GetFor(ids, _termsLocation.ToSpan(), -1L);
+        
+        if (ids.Length < 128)
+            goto SkipPagePrefetching;
```
Why?

maciejaszyk replied:
Removed. I was testing if there are scenario when sorting and preparing data for prefetching makes it slower.

maciejaszyk committed:
RavenDB-22469: Always prefetch pages with terms. Read terms in bulk to avoid extensive memory allocations in the TermsReader cache.

maciejaszyk commented:
https://issues.hibernatingrhinos.com/issue/RavenDB-22469/Slow-deep-paging-in-Corax#focus=Comments-67-1066111.0-0

ayende commented on src/Corax/Querying/TermsReader.cs:
```diff
@@ -41,6 +51,79 @@ public string GetTermFor(long id)
+    {
+        const int pageSizeShift = 13;
+        Debug.Assert(1 << pageSizeShift == (long)Voron.Global.Constants.Storage.PageSize);
+        var maxToProcess = Math.Min(ids.Length, 1024);
```
Is there a reason why you put a limit here? If so, we probably should document here with a comment

arekpalinski replied:
@maciejaszyk ping

arekpalinski replied:
I wonder whether we shouldn't do it outside? So the called would handle the batching explicitly

maciejaszyk replied:
The point is that when dealing with large datasets, we don't want to internally allocate such large arrays. It's better to work with a small cache. I've played with the size and I've not noticed big differences.

maciejaszyk replied:
> I wonder whether we shouldn't do it outside? So the called would handle the batching explicitly

The TermsReader is the cache owner, we do not want upper caller to reserve big chunks. The reader should take care of this instead.

arekpalinski replied:
Okay, so please add a comment explaining the selected batch side so one will know that it's not a random number but selected based on actual testing results

arekpalinski replied:
@maciejaszyk I don't see any comment added here

maciejaszyk replied:
Done

arekpalinski commented on src/Corax/Querying/TermsReader.cs:
```diff
@@ -41,6 +51,79 @@ public string GetTermFor(long id)
+        _rawTermsContainer.Count = ids.Length;
+        Container.GetAll(_llt, _termsLocation.ToSpan(), _rawTermsContainer.RawItems, -1L, _llt.PageLocator);
+        termsSet = _rawTermsContainer.ToSpan();
+        return maxToProcess;
```
Is it okay to return "max" here? Asking about it since I see that we have "and remove duplicates" call above. Do we care about that?

arekpalinski replied:
Okay, I found below that this is for batching purposes only, please check my above comment with idea of moving that to the caller.

maciejaszyk replied:
Context: remove duplicates is used to deduplicate the pageprefetch list.

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs:
```diff
@@ -419,12 +420,18 @@ public long RegisterDuplicates<TProjection>(ref TProjection hasProjection, long
+                    
+                    
+                    // Assumptions: we're in Map, so that mean we have ID of the doc saved in the tree. So we want to keep track what we returns
+                  //  _documentIdReader.GetAllTermsFromSet(distinctIds, out var keys);
```
It this comment in correct place? There is just `return` after it.

Is the commented code intentional?

maciejaszyk replied:
removed

maciejaszyk committed:
RavenDB-22469 Add pagination based on Corax's entry ids for case when it is possible.

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs:
```diff
@@ -419,11 +434,21 @@ public long RegisterDuplicates<TProjection>(ref TProjection hasProjection, long
-                    // Assumptions: we're in Map, so that mean we have ID of the doc saved in the tree. So we want to keep track what we returns
-                    foreach (var id in distinctIds)
+                    var isSorting = _query.Metadata.OrderBy is not null;
+                    if (isSorting || _canPerformPaginationBasedOnEntriesIds == false)
```
Why we're not able to use the more efficient pagination mode when sorting is involved?

arekpalinski replied:
@maciejaszyk ?

maciejaszyk replied:
Removed. EntryId deduplication is done via DeduplicationMatch now.

arekpalinski reviewed (commented):
Please check the comment about `isSorting`

ayende commented on src/Corax/Indexing/IndexWriter.cs:
```diff
@@ -257,13 +278,21 @@ public IndexEntryBuilder Index(ReadOnlySpan<byte> key)
 
+        private void DisablePaginationBasedOnEntryIdSupport()
+        {
+            _indexMetadata.Add(Constants.IndexWriter.PaginationBasedOnEntryIdSupportStatus, (long)EntryIdPaginationSupportStatus.Disabled);
```
Also need to `PaginationBasedOnEntryIdSupportStatus  = Disabled` here, no?

maciejaszyk replied:
yes, added

ayende commented on src/Corax/SupportedFeatures.cs:
```diff
@@ -1,9 +1,10 @@
     
     public readonly bool PhraseQuery = isPhraseQuerySupported;
     public readonly bool StoreOnly = isStoreOnlySupported;
+    public readonly bool PaginationBasedOnEntryId = isPaginationBasedOnEntryIdSupported;
```
When will this _not_ be true?

maciejaszyk replied:
I bumped the index version, so older indexes will have `false` on `IndexWriter` creation.

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs:
```diff
@@ -390,14 +396,23 @@ public void Initialize(Index index, IndexQueryServerSide query, IndexSearcher se
-               
+                
+                _canPerformPaginationBasedOnEntriesIds = searcher.EntryIdPaginationSupportStatus == EntryIdPaginationSupportStatus.Supported 
+                                                        && _isMap
```
This will not work on map/reduce?

maciejaszyk replied:
I tried, however I found a problem even with the previous code: 
https://issues.hibernatingrhinos.com/issue/RavenDB-24794/Incorrect-pagination-via-streaming-query.

I think it is possible, but the issue above has to be fixed first.

maciejaszyk replied:
Fixed in this PR

redknightlois commented on src/Corax/Querying/TermsReader.cs:
```diff
@@ -41,6 +51,79 @@ public string GetTermFor(long id)
+                Vector128.ShiftRightArithmetic(containers, pageSizeShift).Store(ptr);
+            }
+        }
+
```
Did you measure differences here between using 512 bits fallbacks instead of having the code replicated for 256 and 128? As funny as it sounds there are cases that using the fallback is faster than having all the specific code (Microsoft made that mistake with CosineSimilarity).

maciejaszyk replied:
I have not. However, do you remember the perf difference between approaches? CosineSimilarity is used more intensively and performs a lot more operations than this, and I doubt this could give us real benefits.

redknightlois replied:
30% it was big.

ayende replied:
was this checked?

maciejaszyk replied:
https://gist.github.com/maciejaszyk/2f96481988395cbc28cab9396d97ab15

redknightlois replied:
Based on those numbers, I would just do 512; just because of complexity of the function. The extra code is not bringing extra bang.

ayende replied:
And when you don't have 512 (very common) ?

What I don't understand, after JIT, I would expect both function to be exactly the same.
Isn't `AdvInstructionSet.IsAcceleratedVector512` a jit time constant?

arekpalinski replied:
@redknightlois ?

redknightlois replied:
When we don't have 512 it will fallback to use 2 lanes. The execution is the same but it will use 2x256 vectors. It is as if you are writing a 2 times unrolled loop.

redknightlois replied:
I think there is a disconnection in what I mean and what @ayende is asking... 

I would write this code only:

```
            var N = Vector512<long>.Count;
            for (; idX + N < _pagesToPrefetch.Count; idX += N)
            {
                var ptr = _pagesToPrefetch.RawItems + idX;
                var containers = Vector512.Load(ptr);
                Vector512.ShiftRightArithmetic(containers, pageSizeShift).Store(ptr);
            }
```

No checks... just plain code... why? Because that will fallback through the execution based on what length of instructions we have. In 256 it will issue 2x256 instead of a single 512. In 128 it will issue 4x128 instead of a single 512. Which is essentially free unrolling.

ayende replied:
Good idea, yes.

arekpalinski replied:
@maciejaszyk as I see this comment isn't still resolved. Can you please take a look here?

arekpalinski replied:
But after checking the actual code I see the code is written exactly as suggested in the above comment. So I believe the comment wasn't just marked as resolved. 

@maciejaszyk please double check

maciejaszyk replied:
Done.

maciejaszyk committed:
RavenDB-24794 Perform deduplication of corax entries ids inside Corax to ensure proper pagination and correctness of total results.

maciejaszyk committed:
RavenDB-22469 Deduplication in `MoreLikeThis` must be added at the end.

arekpalinski commented on src/Corax/Querying/Matches/Meta/IQueryMatch.cs:
```diff
@@ -77,3 +79,9 @@ public enum SkipSortingResult
     SortingIsRequired
 }
+
+public enum Duplicates
```
Maybe `DuplicatesOccurrence`?

ayende commented on src/Corax/Utils/GrowableHashSet.cs:
```diff
@@ -0,0 +1,74 @@
+
+namespace Corax.Utils;
+
+public sealed class GrowableHashSet<TItem>
```
why do we have this? So we can handle > 2B items to find?

If so, increasing when the count is at `int.MaxValue` is _super_ late, you already hit the max fill factor.

maciejaszyk replied:
Yes, that was the reason why we created it. Previously, it was in Raven.Server, but I moved that code to Corax.

ayende replied:
So it needs to be viable, what if we'll use something like this:
https://gist.github.com/ayende/c0586f194d37db5262f07aca5c7c6341

roughly based on roaring bitmaps

maciejaszyk replied:
@ayende @redknightlois This code is old and well tested. I think changing it to a different data structure would be better done in a different ticket and investigated/benchmarked properly: 
https://issues.hibernatingrhinos.com/issue/RavenDB-24838/Investigate-deduplication-via-bitmap-instead-hashset-in-Corax

redknightlois replied:
Agree

ayende replied:
Note that this is "Well tested" only in the sense that it was in our codebase.
I'm fairly certainly that if you'll try to throw 5B items there, it would fail catastrophically.

ayende replied:
Having said that, handling that in a separate PR is best

ayende commented on src/Corax/Utils/GrowableHashSet.cs:
```diff
@@ -0,0 +1,74 @@
+
+    public bool Add(TItem item)
+    {
+        if (_newestHashSet!.Count >= _maxSizePerCollection)
```
In order to reach here, we already have a 2B backing array, and we have a 100% used it.
So you have a lot of collisions.

ayende replied:
The default should be a max of 512M elements, I think.
Assuming load factor of 70% - that puts us under 1B elements. 
Now we'll need a new one here.

redknightlois replied:
even 512M elements is A LOT!!! Beyond 10M there are better ways to construct things to be able to handle that appropriately.

ayende replied:
In this case, we have a scenario where we _usually_ never hit anything like that.
The problem is what happens when the user does a query over 5B items?

redknightlois replied:
That's what I mean. If we hit that we should be doing it differently.

maciejaszyk replied:
Resolved by introducing bitmap as container. Currently, GrowableHashset will be used only on small sets.

maciejaszyk committed:
RavenDB-22469

- Renamed flag about duplicates in corax
- Added explanation for buffer size in TermsReader
- We can perform pagination based on entries Ids in case of sorting

redknightlois commented on src/Corax/Querying/Matches/DeduplicationMatch.cs:
```diff
@@ -0,0 +1,62 @@
+                var currentId = Unsafe.Add(ref startBuffer, i);
+                Unsafe.Add(ref startBuffer, newResults) = currentId;
+                newResults += _ids.Add(currentId).ToInt32();
+            }
```
Do we have a measurement of how impactful is this loop?

arekpalinski commented:
Please rebase

maciejaszyk committed:
RavenDB-22469 Use a bitmap instead of a growable hashset to reduce GC pressure.

arekpalinski commented:
test this in debug please

maciejaszyk committed:
RavenDB-22469 Implementation of growable bitmap.

ayende commented on src/Corax/Querying/Matches/DeduplicationMatch.cs:
```diff
@@ -0,0 +1,125 @@
+    public DeduplicationMatch(IndexSearcher indexSearcher, TInner inner, bool forceHashSet)
+    {
+        long maxBitId = indexSearcher.LastEntryId;
+        long numberOfEntries = indexSearcher.NumberOfEntries; // ensure not zero
```
how do you ensure not zero?

maciejaszyk replied:
That's actually a leftover comment from development. I made BitArray inclusive (added +1 to the requested capacity) on the accepting range for simplicity, instead of decrementing entryId by 1.

ayende commented on src/Corax/Querying/Matches/DeduplicationMatch.cs:
```diff
@@ -0,0 +1,125 @@
+    {
+        long maxBitId = indexSearcher.LastEntryId;
+        long numberOfEntries = indexSearcher.NumberOfEntries; // ensure not zero
+        var bitmapMemoryRequiredInBytes = maxBitId / 64;
```
why is that 64 and count in bytes?
doesn't this count in _longs_ ?

maciejaszyk replied:
Yes, fixed.

ayende commented on src/Corax/Querying/Matches/DeduplicationMatch.cs:
```diff
@@ -0,0 +1,125 @@
+
+        // If the bitmap is big enough and actually only around 1.5% entries exist, we will use a HashSet instead.
+        if (bitmapMemoryRequiredInBytes > IndexSearcher.BitmapMemoryRequiredThresholdInBytes 
+            && entriesMemoryRequiredInBytes < (bitmapMemoryRequiredInBytes >> 6) || forceHashSet)
```
how do you get from `>>6` to 1.5%?
Better to just do a `/64` which is much more readable

ayende replied:
Also add here that the total number of elements is < 1G

maciejaszyk replied:
Done

redknightlois replied:
for reference division by constant gets converted either to multiplications or shifts (whatever is faster) by the JIT.  and the same for multiplications if they can be converted to sums and shifts.

maciejaszyk replied:
> Also add here that the total number of elements is < 1G

@ayende can you elaborate more about that condition?

ayende replied:
If we have 1 billion elements, what would the growable hash set do? Probably fail badly, no?

maciejaszyk replied:
I’d remove the number-of-entries condition entirely. With low density and 1B total entries, it would switch to a hash set, and the max matches would be 15M—well within what that structure can handle efficiently. Otherwise, we will create a bitarray. What do you think?

ayende replied:
yes, go ahead and do that.

maciejaszyk replied:
done

ayende commented on src/Corax/Querying/Matches/DeduplicationMatch.cs:
```diff
@@ -0,0 +1,125 @@
+                Unsafe.Add(ref startBuffer, newResults) = currentId;
+                newResults += match._bitmap.Add(currentId).ToInt32();
+            }
+        } while (read > 0 && newResults == 0);
```
Why do you have `newResults == 0` here? That is if I filtered out everything, no?
But shouldn't I try to read as much as possible before returning ?

maciejaszyk replied:
It depends. Duplicates are not something that happen a lot, and sometimes we don't need an exact result count from a query. I thought that part would be better handled by RavenDB part.

ayende replied:
Add a comment to that effect, to explain this.

maciejaszyk replied:
added

ayende commented on src/Corax/Querying/Matches/DeduplicationMatch.cs:
```diff
@@ -0,0 +1,125 @@
+                Unsafe.Add(ref startBuffer, newResults) = currentId;
+                newResults += match._hashset.Add(currentId).ToInt32();
+            }
+        } while (read > 0 && newResults == 0);
```
same

maciejaszyk replied:
added

ayende commented on src/Corax/Utils/GrowableHashSet.cs:
```diff
@@ -0,0 +1,74 @@
+    {
+        _comparer = comparer;
+        _hashSetsBucket = null;
+        _maxSizePerCollection = maxSizePerCollection ?? int.MaxValue;
```
This needs to change to max of 128M

maciejaszyk replied:
done

maciejaszyk committed:
RavenDB-22469 Added new thresholds | Bug fix

redknightlois commented on test/SlowTests/Corax/GrowableBitArrayTests.cs:
```diff
@@ -0,0 +1,69 @@
+        Assert.False(lookup.Add(GrowableBitArray.MaxCapacityPerBitmapInBits+2));
+        Assert.Throws<ArgumentOutOfRangeException>(() => lookup.Add(GrowableBitArray.MaxCapacityPerBitmapInBits+3));
+    }
+}
```
We should add a fuzzying test here too. Fuzzy construction and fuzzy modifications.

maciejaszyk replied:
added

arekpalinski commented on test/SlowTests/Corax/GrowableBitArrayTests.cs:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+
+    [Fact]
```
RavenFact

maciejaszyk replied:
fixed

arekpalinski commented on src/Corax/Utils/GrowableBitArray.cs:
```diff
@@ -0,0 +1,98 @@
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public bool Add(long id)
+    {
+        if (id > _capacity)
```
Shouldn't be `>=`?

redknightlois replied:
Bit Arrays as we built them originally are 0 based (you could address the 0-bit). So yes `==` will be over capacity, great catch. I would suggest to add a test to ensure that is the case before fixing it.

maciejaszyk replied:
Yes. Fixed.

arekpalinski commented on src/Corax/Utils/GrowableBitArray.cs:
```diff
@@ -0,0 +1,98 @@
+    public bool Add(long id)
+    {
+        if (id > _capacity)
+            throw new ArgumentOutOfRangeException($"Tried to modify bit no {id}, however the capacity is only { _capacity * 64}");
```
"bit no"?

maciejaszyk committed:
RavenDB-22469 Bug fixes | PR comments

arekpalinski commented:
@maciejaszyk I approved the PR but I want to go over all comments with you together before the merge

arekpalinski reviewed (approved):
The PR is ready but please wait for @maciejaszyk's explicit approval before merging it. He still wants to double-check something here.

maciejaszyk commented:
It's ready to merge.
